src/Npgsql/NpgsqlCommand.cs

```diff
@@ -1120,116 +1150,182 @@ public new Task<NpgsqlDataReader> ExecuteReaderAsync(CommandBehavior behavior, C
+                    {
+                        // The waiting on the ExecutionCompletion ManualResetValueTaskSource is necessarily
+                        // asynchronous, so allowing sync would mean sync-over-async.
+                        throw new NotSupportedException(
+                            "Synchronous command execution is not supported when multiplexing is on");
+                    }
+
+                    ValidateParameters(conn.Pool!.MultiplexingTypeMapper!);
+                    ProcessRawQuery();
 
                     State = CommandState.InProgress;
 
-                    if (Log.IsEnabled(NpgsqlLogLevel.Debug))
-                        LogCommand(connector.Id);
-                    NpgsqlEventSource.Log.CommandStart(CommandText);
-                    Task sendTask;
-
-                    // If a cancellation is in progress, wait for it to "complete" before proceeding (#615)
-                    lock (connector.CancelLock) { }
-
-                    connector.UserTimeout = CommandTimeout * 1000;
-
-                    // We do not wait for the entire send to complete before proceeding to reading -
-                    // the sending continues in parallel with the user's reading. Waiting for the
-                    // entire send to complete would trigger a deadlock for multi-statement commands,
-                    // where PostgreSQL sends large results for the first statement, while we're sending large
-                    // parameter data for the second. See #641.
-                    // Instead, all sends for non-first statements and for non-first buffers are performed
-                    // asynchronously (even if the user requested sync), in a special synchronization context
-                    // to prevents a dependency on the thread pool (which would also trigger deadlocks).
-                    // The WriteBuffer notifies this command when the first buffer flush occurs, so that the
-                    // send functions can switch to the special async mode when needed.
-                    sendTask = (behavior & CommandBehavior.SchemaOnly) == 0
-                        ? SendExecute(connector, async)
-                        : SendExecuteSchemaOnly(connector, async);
-
-                    // The following is a hack. It raises an exception if one was thrown in the first phases
-                    // of the send (i.e. in parts of the send that executed synchronously). Exceptions may
-                    // still happen later and aren't properly handled. See #1323.
-                    if (sendTask.IsFaulted)
-                        sendTask.GetAwaiter().GetResult();
+                    // TODO: Experiment: do we want to wait on *writing* here, or on *reading*?
+                    // Previous behavior was to wait on reading, which throw the exception from ExecuteReader (and not from
+                    // the first read). But waiting on writing would allow us to do sync writing and async reading.
```

Comment:
That bit about doing sync writing is now stale - multiplexing always uses async APIs to write (though they'll usually complete synchronously). Because of this - and because executing a command in multiplexing means enqueuing the a channel, which is also async-only - calling sync Execute APIs on NpgsqlCommand throws NotSupportedException.

There could still be value in cleanly splitting writing and reading across different APIs. This would be a breaking change though, as PG errors are currently raised from NpgsqlDbCommand.Execute*, but if we make this change they'll be raised from NpgsqlDbDataReader.Read instead. Not too sure about this.

Reply:
Breaking change because the trace changes?

Reply:
Breaking change because servers errors no longer get raised from ExecuteReaderAsync. It doesn't necessarily mean we shouldn't do it... I really like the idea of separating write APIs from read APIs, but ADO.NET isn't exactly architected that way :/